Build failure: curl with `http3Support` enabled

### Steps To Reproduce

Steps to reproduce the behavior:

1. check out nixpkgs @ b0d36bd0a420ecee3bc916c91886caca87c894e9 (latest master as of writing)
2. `nix build --impure --expr 'with import ./. {}; pkgs.curl.override { http3Support = true; }' -L`

### Build log

```
curl> In file included from vquic/curl_ngtcp2.c:36:
curl> /nix/store/b4shqdha5xymiz83ibwafgrj894f481j-ngtcp2-1.1.0-dev/include/ngtcp2/ngtcp2_crypto_quictls.h:73:5: error: unknown type name 'OSSL_ENCRYPTION_LEVEL'
curl>    73 |     OSSL_ENCRYPTION_LEVEL ossl_level);
curl>       |     ^~~~~~~~~~~~~~~~~~~~~
curl> /nix/store/b4shqdha5xymiz83ibwafgrj894f481j-ngtcp2-1.1.0-dev/include/ngtcp2/ngtcp2_crypto_quictls.h:82:15: error: unknown type name 'OSSL_ENCRYPTION_LEVEL'
curl>    82 | NGTCP2_EXTERN OSSL_ENCRYPTION_LEVEL
curl>       |               ^~~~~~~~~~~~~~~~~~~~~
curl> vquic/curl_ngtcp2.c: In function 'quic_init_ssl':
curl> vquic/curl_ngtcp2.c:524:3: error: implicit declaration of function 'SSL_set_quic_use_legacy_codepoint' [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Werror=implicit-function-declaration8;;]
curl>   524 |   SSL_set_quic_use_legacy_codepoint(ctx->ssl, 0);
curl>       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
curl> cc1: some warnings being treated as errors
curl> make[2]: *** [Makefile:3253: vquic/libcurl_la-curl_ngtcp2.lo] Error 1
curl> make[2]: *** Waiting for unfinished jobs....
curl> make[2]: Leaving directory '/build/curl-8.4.0/lib'
curl> make[1]: *** [Makefile:1503: all] Error 2
curl> make[1]: Leaving directory '/build/curl-8.4.0/lib'
curl> make: *** [Makefile:1273: all-recursive] Error 1
error: builder for '/nix/store/hyvn564y6n9mssn7hvzk6lyn6n2cprg6-curl-8.4.0.drv' failed with exit code 2;
       last 10 log lines:
       > vquic/curl_ngtcp2.c:524:3: error: implicit declaration of function 'SSL_set_quic_use_legacy_codepoint' [-Werror=implicit-function-declaration]
       >   524 |   SSL_set_quic_use_legacy_codepoint(ctx->ssl, 0);
       >       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > cc1: some warnings being treated as errors
       > make[2]: *** [Makefile:3253: vquic/libcurl_la-curl_ngtcp2.lo] Error 1
       > make[2]: *** Waiting for unfinished jobs....
       > make[2]: Leaving directory '/build/curl-8.4.0/lib'
       > make[1]: *** [Makefile:1503: all] Error 2
       > make[1]: Leaving directory '/build/curl-8.4.0/lib'
       > make: *** [Makefile:1273: all-recursive] Error 1
       For full logs, run 'nix log /nix/store/hyvn564y6n9mssn7hvzk6lyn6n2cprg6-curl-8.4.0.drv'.

```

### Notify maintainers

@lovek323 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.69-xanmod1, NixOS, 24.05 (Uakari), 24.05.20231227.cfc3698`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/store/lf2dq57rqn8d03ybrxxi6l8gg27a4av0-source`

```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/cu/curlMinimal/package.nix
+++ b/pkgs/by-name/cu/curlMinimal/package.nix
@@ -14,7 +14,7 @@
     !(isDarwin && (stdenv.buildPlatform != stdenv.hostPlatform))
   ), libkrb5
 , http2Support ? true, nghttp2
-, http3Support ? false, nghttp3, ngtcp2
+, http3Support ? false, nghttp3, ngtcp2, quictls
 , websocketSupport ? false
 , idnSupport ? false, libidn2
 , ldapSupport ? false, openldap
@@ -46,6 +46,10 @@
 # files.
 
 assert !((lib.count (x: x) [ gnutlsSupport opensslSupport wolfsslSupport rustlsSupport ]) > 1);
+
+let
+  openssl' = if http3Support then quictls else openssl;
+in
 
 stdenv.mkDerivation (finalAttrs: {
   pname = "curl";
@@ -106,7 +110,7 @@ stdenv.mkDerivation (finalAttrs: {
     lib.optionals http3Support [ nghttp3 ngtcp2 ] ++
     lib.optional idnSupport libidn2 ++
     lib.optional ldapSupport openldap ++
-    lib.optional opensslSupport openssl ++
+    lib.optional opensslSupport openssl' ++
     lib.optional pslSupport libpsl ++
     lib.optional rtmpSupport rtmpdump ++
     lib.optional scpSupport libssh2 ++
@@ -150,7 +154,7 @@ stdenv.mkDerivation (finalAttrs: {
       (lib.withFeatureAs brotliSupport "brotli" (lib.getDev brotli))
       (lib.withFeatureAs gnutlsSupport "gnutls" (lib.getDev gnutls))
       (lib.withFeatureAs idnSupport "libidn2" (lib.getDev libidn2))
-      (lib.withFeatureAs opensslSupport "openssl" (lib.getDev openssl))
+      (lib.withFeatureAs opensslSupport "openssl" (lib.getDev openssl'))
       (lib.withFeatureAs scpSupport "libssh2" (lib.getDev libssh2))
       (lib.withFeatureAs wolfsslSupport "wolfssl" (lib.getDev wolfssl))
     ]
@@ -205,7 +209,8 @@ stdenv.mkDerivation (finalAttrs: {
   passthru = let
     useThisCurl = attr: attr.override { curl = finalAttrs.finalPackage; };
   in {
-    inherit opensslSupport openssl;
+    inherit opensslSupport;
+    openssl = openssl';
     tests = {
       withCheck = finalAttrs.finalPackage.overrideAttrs (_: { doCheck = true; });
       fetchpatch = tests.fetchpatch.simple.override { fetchpatch = (fetchpatch.override { fetchurl = useThisCurl fetchurl; }) // { version = 1; }; };

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -3055,7 +3055,6 @@ with pkgs;
   };
 
   curlHTTP3 = curl.override {
-    openssl = quictls;
     http3Support = true;
   };
 

